Define build target for mkldnn lib build to fix 'make clean USE_MKMLDNN=1' issue

jinhuang415 opened the pull request:
This is a follow up for https://github.com/apache/incubator-mxnet/pull/10783, the view is messed up in that PR so open this new one.

This is to fix the 'make clean USE_MKMLDNN=1' will download and build MKLDNN library issue. As suggested in https://github.com/apache/incubator-mxnet/pull/10783, the new solution is to define a mkldnn make target and integrate into current make dependency, this solution will be cleaner.  It also simplifies current mkldnn library build logic a little bit. One simplification is assume MKLDNNROOT and MKLROOT now share the same directory since we always download mklml and copy the headers/libs to MKLDNN lib directory. Since we used mkldnn.mk to handle build logic,  prepare_mkldnn.sh/prepare_mkl.sh is not needed anymore.

jinhuang415 commented:
@szha @zheng-da I opened a new PR to fix 'make clean USE_MKLDNN=1" issue since previous PR is messed up (https://github.com/apache/incubator-mxnet/pull/10783), please help to take a look at the new solution.

@TaoLv @ashokei @pengzhao-intel I made some change/cleanup about current mkldnn library build logic so there will be less corner cases, would you also help to review the change?

ashokei commented:
@jinhuang415 most of the changes look ok for mkldnn, but i notice you are removing prepare_mkl script too. This script could be used independently of mkldnn (for eg: USE_BLAS=mkl after running prepare_mkl.sh , and build mxnet without USE_MKLDNN should be possible? )

jinhuang415 commented:
@ashokei Thanks for helping review. Currently if USE_BLAS=mkl, libmxnet.so will static link with full MKL library instead of mklml library, and if USE_MKLDNN=1 during mkldnn library build it will use the prepare_mkl.sh under mkldnn repo, so looks prepare_mkl.sh (the one in the mxnet repo root directory) is not needed anymore. Do you think we still need to keep it?

ashokei commented:
@jinhuang415 If a user does not want to build with MKLDNN , but they still want to use mklml lib for BLAS/GEMM apis (USE_BLAS=mkl) , the prepare_mkl.sh script will help. 

If it USE_BLAS=mkl is not working with prepare_mkl.sh script in mxnet repo, we need to fix it.

szha commented:
@ashokei this procedure isn't documented anywhere, so whoever is aware of this approach would be able to just grab the mklml library from mkldnn release page.

szha commented on mkldnn.mk:
```diff
@@ -0,0 +1,66 @@
+# under the License.
+
+ifeq ($(USE_MKLDNN), 1)
+	MKLDNN_ROOTDIR = $(ROOTDIR)/3rdparty/mkldnn
```
it would be nice if you could further eliminate some duplicate variables. also, why override the value here?

jinhuang415 replied:
Actually there is no duplicate variables here, MKLDNN_ROOTDIR is the mkldnn submodule root directory under 3rdparty/mkldnn, and other variables (like MKLDNN_BUILDDIR/MKLDNN_INSTALLDIR) are based on this variable. And MKLDNNROOT is the variable to point to the final generated library/header file root which will be used by mxnet directly. So no override here.

szha replied:
Can the MKLDNNROOT be renamed to avoid confusion? The goal is that you don't have to explain this to the next 100 people who look at this file.

szha commented:
Ping @jinhuang415

jinhuang415 commented:
@szha We are under some internal discussion on how to make the change, will update later.

jinhuang415 commented:
@szha After discussion we still keep current change, for mkldnn build to keep simple we always download mklml small library, for mxnet build if blas is mkl we will use full mkl as blas (so prepare_mkl.sh is not necessary). I also fixed an issue that mxnet will link both mklml and full mkl when MKLDNN is enabled and blas is mkl (the expected behavior should link with full mkl alone). Please help to take a look.

jinhuang415 commented:
@szha We removed MKLDNN_ROOT (make variable) support in new diff since we found it is not necessary to use another directory to hold the built library/header file(we can always use 3rdparty/mkldnn/install), if there is need to use pre-installed mkldnn library we can add support in another PR.   Regarding the name, I still use MKLDNNROOT to keep consistent with mkldnn github repo README (https://github.com/intel/mkl-dnn#linking-your-application) and changed some variable names in mkldnn.mk, hope it can make less confusion.

jinhuang415 commented:
@szha Jenkins finally passed after I did re-trigger for 5 times, would you help to check if you have more comments?

zheng-da commented on Makefile:
```diff
@@ -427,11 +426,11 @@ endif
 	$(CXX) -std=c++11 -c $(CFLAGS) -MMD -c $< -o $@
 
-build/src/%_gpu.o: src/%.cu
+build/src/%_gpu.o: src/%.cu | mkldnn
```
does the GPU code depend on MKLDNN?

jinhuang415 replied:
GPU code need ndarray.h which depends on mkldnn (need mkldnn.hpp) when MKLDNN is enabled. The mkldnn target will be empty if MKLDNN is not enabled.

zheng-da commented on Makefile:
```diff
@@ -427,11 +426,11 @@ endif
 	$(CXX) -std=c++11 -c $(CFLAGS) -MMD -c $< -o $@
 
-build/src/%_gpu.o: src/%.cu
+build/src/%_gpu.o: src/%.cu | mkldnn
```
does the GPU code depend on MKLDNN?

jinhuang415 replied:
GPU code need ndarray.h which depends on mkldnn (need mkldnn.hpp) when MKLDNN is enabled. The mkldnn target will be empty if MKLDNN is not enabled.
